ql/src/java/org/apache/hadoop/hive/ql/optimizer/BucketVersionPopulator.java

```diff
@@ -131,22 +132,47 @@ private OpGroup newGroup(NodeProcessorCtx procCtx) {
     }
   }
 
+  enum InfoType {
+    MANDATORY, OPTIONAL,
+  };
+
   /**
    * This class represents the version required by an Operator.
    */
-  private static class OperatorBucketingVersionInfo {
+  static class OperatorBucketingVersionInfo {
+
+    public static final Comparator<OperatorBucketingVersionInfo> MANDATORY_FIRST =
+        new Comparator<BucketVersionPopulator.OperatorBucketingVersionInfo>() {
 
+          @Override
+          public int compare(OperatorBucketingVersionInfo i1, OperatorBucketingVersionInfo i2) {
+            int r = i1.infoType.compareTo(i2.infoType);
+            if (r != 0) {
+              // mandatory first
+              return r;
+            }
+            r = i2.bucketingVersion - i1.bucketingVersion;
```

Comment:
nit. `Integer.compare(i2.bucketingVersion, i1.bucketingVersion)` can be used here.

Reply:
:+1: